# Data Analysis

In [1]:
import pandas as pd
import numpy as np

In [2]:
survey = pd.read_csv('data/surveys_responses.csv')
survey

TEXT1  \
0    HABEAS CORPUS. AÇÃO DE DESTITUIÇÃO DO PODER FA...   
1    HABEAS CORPUS. AÇÃO DE DESTITUIÇÃO DO PODER FA...   
2    PROCESSUAL CIVIL. AGRAVO REGIMENTAL. ART. 535 ...   
3    PROCESSUAL CIVIL. AGRAVO REGIMENTAL. ART. 535 ...   
4    AGRAVO  REGIMENTAL  NO  HABEAS  CORPUS. DIREIT...   
..                                                 ...   
235  AGRAVO REGIMENTAL EM RECURSO ESPECIAL. SEGURO....   
236  MANDADO DE SEGURANÇA. LEASING. ISS. AUTORIDADE...   
237  MANDADO DE SEGURANÇA. LEASING. ISS. AUTORIDADE...   
238  MANDADO DE SEGURANÇA. ADMINISTRATIVO. POLICIAL...   
239  MANDADO DE SEGURANÇA. ADMINISTRATIVO. POLICIAL...   

                                                 TEXT2  SIMILARITY_SCORE  \
0    RECURSO ESPECIAL. AÇÃO INDENIZATÓRIA. LEVANTAM...                 0   
1    RECURSO ESPECIAL. AÇÃO INDENIZATÓRIA. LEVANTAM...                 0   
2    CIVIL E CONSUMIDOR. RECURSO ESPECIAL. CONTRATO...                 1   
3    CIVIL E CONSUMIDOR. RECURSO ESPECIAL. CONTRATO...                 2   
4    PROCESSUAL  PENAL.  CRIME  DE TRÂNSITO. DIREÇÃ...                 2   
..                                                 ...               ...   
235  RECURSO ESPECIAL. PENAL. TRÁFICO ILÍCITO DE DR...                 0   
236  PROCESSUAL CIVIL. ADMINISTRATIVO. MANDADO DE S...                 0   
237  PROCESSUAL CIVIL. ADMINISTRATIVO. MANDADO DE S...                 0   
238  ADMINISTRATIVO.  MANDADO  DE  SEGURANÇA.  SERV...                 3   
239  ADMINISTRATIVO.  MANDADO  DE  SEGURANÇA.  SERV...                 1   

     SIMILARITY_CONFIDENCE                                    TEXT1_HIGHLIGHT  \
0                        4  A observância do cadastro de adotantes, não é ...   
1                        4                                    Não relacionado   
2                        4   Inviável a análise de matéria federal não pre...   
3                        4  O acórdão recorrido está expressamente embasad...   
4                        4                os  crimes de embriaguez ao volante   
..                     ...                                                ...   
235                      4  AGRAVO REGIMENTAL EM RECURSO ESPECIAL. SEGURO....   
236                      0                                   sem similaridade   
237                      4  MANDADO DE SEGURANÇA. LEASING. ISS. AUTORIDADE...   
238                      1                               mandado de segurança   
239                      3  1. O impetrante foi demitido por ter se utiliz...   

    TEXT1_HIGHLIGHT_POSITION  \
0                      Corpo   
1                      Corpo   
2                      Corpo   
3                      Corpo   
4                      Corpo   
..                       ...   
235                Cabeçalho   
236                    Corpo   
237                Cabeçalho   
238                    Corpo   
239                    Corpo   

                                       TEXT2_HIGHLIGHT  \
0     ausência de impugnação dos fundamentos do acó...   
1                                      Não relacionado   
2    deve-se reconhecer a violação ao art. 51, I, d...   
3    Anulada a cláusula que limita o valor da inden...   
4    dirigir veículo automotor em via\npública  com...   
..                                                 ...   
235  RECURSO ESPECIAL. PENAL. TRÁFICO ILÍCITO DE DR...   
236                                   sem similaridade   
237  PROCESSUAL CIVIL. ADMINISTRATIVO. MANDADO DE S...   
238                               mandado de segurança   
239  1.  Mandado  de  segurança  contra  ato do Sr....   

    TEXT2_HIGHLIGHT_POSITION  
0                      Corpo  
1                      Corpo  
2                      Corpo  
3                      Corpo  
4                      Corpo  
..                       ...  
235                Cabeçalho  
236                    Corpo  
237                Cabeçalho  
238                    Corpo  
239                    Corpo  

[240 rows x

### Pares com mais de um rótulo

In [7]:
survey_one_label = survey[survey[['TEXT1','TEXT2','SIMILARITY_SCORE']].groupby(['TEXT1','TEXT2'])['SIMILARITY_SCORE'].transform('count') == 1]
survey_one_label.to_csv('data/one_label.csv')
len(survey_one_label)

40

In [4]:
survey_two_label = survey[survey[['TEXT1','TEXT2','SIMILARITY_SCORE']].groupby(['TEXT1','TEXT2'])['SIMILARITY_SCORE'].transform('count') > 1]
survey_two_label

TEXT1  \
0    HABEAS CORPUS. AÇÃO DE DESTITUIÇÃO DO PODER FA...   
1    HABEAS CORPUS. AÇÃO DE DESTITUIÇÃO DO PODER FA...   
2    PROCESSUAL CIVIL. AGRAVO REGIMENTAL. ART. 535 ...   
3    PROCESSUAL CIVIL. AGRAVO REGIMENTAL. ART. 535 ...   
4    AGRAVO  REGIMENTAL  NO  HABEAS  CORPUS. DIREIT...   
..                                                 ...   
235  AGRAVO REGIMENTAL EM RECURSO ESPECIAL. SEGURO....   
236  MANDADO DE SEGURANÇA. LEASING. ISS. AUTORIDADE...   
237  MANDADO DE SEGURANÇA. LEASING. ISS. AUTORIDADE...   
238  MANDADO DE SEGURANÇA. ADMINISTRATIVO. POLICIAL...   
239  MANDADO DE SEGURANÇA. ADMINISTRATIVO. POLICIAL...   

                                                 TEXT2  SIMILARITY_SCORE  \
0    RECURSO ESPECIAL. AÇÃO INDENIZATÓRIA. LEVANTAM...                 0   
1    RECURSO ESPECIAL. AÇÃO INDENIZATÓRIA. LEVANTAM...                 0   
2    CIVIL E CONSUMIDOR. RECURSO ESPECIAL. CONTRATO...                 1   
3    CIVIL E CONSUMIDOR. RECURSO ESPECIAL. CONTRATO...                 2   
4    PROCESSUAL  PENAL.  CRIME  DE TRÂNSITO. DIREÇÃ...                 2   
..                                                 ...               ...   
235  RECURSO ESPECIAL. PENAL. TRÁFICO ILÍCITO DE DR...                 0   
236  PROCESSUAL CIVIL. ADMINISTRATIVO. MANDADO DE S...                 0   
237  PROCESSUAL CIVIL. ADMINISTRATIVO. MANDADO DE S...                 0   
238  ADMINISTRATIVO.  MANDADO  DE  SEGURANÇA.  SERV...                 3   
239  ADMINISTRATIVO.  MANDADO  DE  SEGURANÇA.  SERV...                 1   

     SIMILARITY_CONFIDENCE                                    TEXT1_HIGHLIGHT  \
0                        4  A observância do cadastro de adotantes, não é ...   
1                        4                                    Não relacionado   
2                        4   Inviável a análise de matéria federal não pre...   
3                        4  O acórdão recorrido está expressamente embasad...   
4                        4                os  crimes de embriaguez ao volante   
..                     ...                                                ...   
235                      4  AGRAVO REGIMENTAL EM RECURSO ESPECIAL. SEGURO....   
236                      0                                   sem similaridade   
237                      4  MANDADO DE SEGURANÇA. LEASING. ISS. AUTORIDADE...   
238                      1                               mandado de segurança   
239                      3  1. O impetrante foi demitido por ter se utiliz...   

    TEXT1_HIGHLIGHT_POSITION  \
0                      Corpo   
1                      Corpo   
2                      Corpo   
3                      Corpo   
4                      Corpo   
..                       ...   
235                Cabeçalho   
236                    Corpo   
237                Cabeçalho   
238                    Corpo   
239                    Corpo   

                                       TEXT2_HIGHLIGHT  \
0     ausência de impugnação dos fundamentos do acó...   
1                                      Não relacionado   
2    deve-se reconhecer a violação ao art. 51, I, d...   
3    Anulada a cláusula que limita o valor da inden...   
4    dirigir veículo automotor em via\npública  com...   
..                                                 ...   
235  RECURSO ESPECIAL. PENAL. TRÁFICO ILÍCITO DE DR...   
236                                   sem similaridade   
237  PROCESSUAL CIVIL. ADMINISTRATIVO. MANDADO DE S...   
238                               mandado de segurança   
239  1.  Mandado  de  segurança  contra  ato do Sr....   

    TEXT2_HIGHLIGHT_POSITION  
0                      Corpo  
1                      Corpo  
2                      Corpo  
3                      Corpo  
4                      Corpo  
..                       ...  
235                Cabeçalho  
236                    Corpo  
237                Cabeçalho  
238                    Corpo  
239                    Corpo  

[200 rows x

#### Então dos 140 pares que deveriam ser anotados por 2 anotadores, 40 pares foram anotados por apenas 1.

In [5]:
heuristic = pd.read_csv('data/to_anotate.csv')
heuristic

sentence_A  \
0    PROCESSUAL CIVIL E ADMINISTRATIVO. SERVIDÃO AD...   
1    AGRAVO  INTERNO  NO  AGRAVO  EM  RECURSO  ESPE...   
2    HABEAS CORPUS. AÇÃO DE DESTITUIÇÃO DO PODER FA...   
3    PROCESSUAL CIVIL. AÇÃO CIVIL PÚBLICA. AMBIENTA...   
4    AGRAVO REGIMENTAL. RECURSO ORDINÁRIO. MANDADO ...   
..                                                 ...   
136  AGRAVO REGIMENTAL NO AGRAVO EM RECURSO ESPECIA...   
137  AGRAVO REGIMENTAL NO AGRAVO EM RECURSO ESPECIA...   
138  PROCESSUAL CIVIL E ADMINISTRATIVO. AGRAVO REGI...   
139  AGRAVO REGIMENTAL NO RECURSO EM MANDADO DE SEG...   
140  ADMINISTRATIVO.   AMBIENTAL.  AÇÃO  CIVIL  PÚB...   

                                            sentence_B  score  
0    PROCESSUAL CIVIL E ADMINISTRATIVO - DESAPROPRI...      2  
1    AGRAVO  INTERNO NO AGRAVO EM RECURSO ESPECIAL....      4  
2    RECURSO  ESPECIAL. PORTE DE ARMA DE FOGO E MUN...      0  
3    DIREITO CIVIL. FAMÍLIA. ADOÇÃO DE MENORES POR ...      0  
4    ADMINISTRATIVO. RECURSO ORDINÁRIO EM MANDADO D...      4  
..                                                 ...    ...  
136  AGRAVO REGIMENTAL NO AGRAVO EM RECURSO ESPECIA...      2  
137  PROCESSUAL CIVIL. AGRAVO REGIMENTAL NO AGRAVO ...      4  
138  PROCESSUAL CIVIL - ADMINISTRATIVO - RECURSO ES...      4  
139  ADMINISTRATIVO  E  PROCESSUAL  CIVIL.  MANDADO...      2  
140  AGRAVO REGIMENTAL EM RECURSO ESPECIAL - AÇÃO D...      0  

[141 rows x 3 columns]

## Questões de Pesquisa

1. Quão relacionadas estão os rótulos dos especialistas para um mesmo par de documentos?
2. Onde estão concentrados os trechos principais dos documentos, corpo ou cabeçalho?
3. Qual a distribuição dos rótulos no conjunto de dados anotado heuristicamente?
4. Qual a distribuição dos rótulos no conjunto de dados anotado pelos especialistas?
5. Quão relacionados estão os rótulos dos especialistas e da heurística?
6. Qual a média e desvio padrão da confiança dos especialistas aos rótulos atribuídos?

In [8]:
# https://stackoverflow.com/questions/15222754/groupby-pandas-dataframe-and-select-most-common-value
survey_tmp = survey_two_label[['TEXT1','TEXT2', 'SIMILARITY_SCORE']].astype('object').groupby(['TEXT1','TEXT2'])['SIMILARITY_SCORE']\
                                                    .agg(pd.Series.mode).to_frame().reset_index()
survey_tmp

ValueError: Must produce aggregated value

In [ ]:
survey_discrepancy = survey_tmp.iloc[[x.Index for x in survey_tmp.itertuples() if type(x.SIMILARITY_SCORE)== np.ndarray]]
survey_discrepancy.to_csv('data/discrepancy.csv')
survey_discrepancy

In [ ]:
survey_agreement = survey.iloc[[x.Index for x in survey_tmp.itertuples() if type(x.SIMILARITY_SCORE) == int]]
survey_agreement

### 1. Quão relacionadas estão os rótulos dos especialistas para um mesmo par de documentos?

- Inicialmente foram usados 120 pares de documentos diferentes, para se obter 2 rótulos por par, objetivando-se 240 registros. No entanto, verificamos que 20 pares dos 120 pares de documentos foram anotados por apenas 1 anotador, deste modo apenas 100 pares foram rótulados por mais de um anotador. No final obteve-se 

Dos 100 pares anotados por mais de um anotador, 32 anotadores concordaram com o mesmo rótulo, enquanto 68 rotularam de maneiras distintas.
A variância e o desvio padrão entre os 68 pares rotulados distintamente são os seguintes:

In [ ]:
len(survey_agreement)

In [ ]:
survey_agreement['SIMILARITY_SCORE'].value_counts().sort_values().plot(kind = 'bar', xlabel='Rótulos', ylabel='Quantidade')

In [ ]:
tmp = np.array([t.tolist() for t in survey_discrepancy.SIMILARITY_SCORE.values])
mean_variance = np.mean(np.var(tmp, axis=1))
mean_std = np.std(np.std(tmp, axis=1))
print('Mean variance: ', mean_variance, ' Mean STD: ', mean_std)